src/main/perf/DocGrouper.java

```diff
@@ -0,0 +1,438 @@
+    }
+
+    public DocGroups getNextDocGroups() throws InterruptedException {
+        return outputQueue.take();
+    }
+
+    public static void initGroupIds(Random random) {
+        assert group100 == null;
+        group100 = randomStrings(100, random);
+        group10K = randomStrings(10000, random);
+        group100K = randomStrings(100000, random);
+        group1M = randomStrings(1000000, random);
+        group2M = randomStrings(2000000, random);
+    }
+
+    static int getTargetNumGroups(int numDocs) {
+        if (numDocs >= 5000000) {
+            return 1_000_000;
+        } else if (numDocs >= 500000) {
+            return 100_000;
+        } else {
+            return 10_000;
+        }
+    }
+
+    // returned array will not have dups
+    private static BytesRef[] randomStrings(int count, Random random) {
+        final BytesRef[] strings = new BytesRef[count];
+        HashSet<String> idSet = new HashSet<>(count);
+        int i = 0;
+        while(i < count) {
+            String s = randomRealisticUnicodeString(random);
+            while (s.equals("") == false && idSet.contains(s)) {
+                s = randomRealisticUnicodeString(random);
+            }
+            strings[i++] = new BytesRef(s);
+            idSet.add(s);
+        }
+
+        return strings;
+    }
+
+    public abstract void add(LineFileDocs.LineFileDoc lfd) throws InterruptedException;
+
+    /**
+     * A simple impl when we do not need grouping
+     */
+    static final class NoGroupImpl extends DocGrouper {
+
+        NoGroupImpl(int numDocs) {
+            super(numDocs, 0);
+        }
+
+        @Override
+        public void add(LineFileDocs.LineFileDoc lfd) throws InterruptedException {
+            if (lfd == LineFileDocs.END) {
+                outputQueue.put(END);
+            }
+            outputQueue.put(new DocGroups.SingleLFD(lfd));
+        }
+    }
+
+    static final class TextGrouper extends DocGrouper {
+
+        private int groupCounter;
+        private int docCounter;
+        private int nextNumDocs;
+        private LineFileDocs.LineFileDoc[] buffer;
+        private final BytesRef[] groupIds;
+        private final float docsPerGroupBlock;
+
+        TextGrouper(int numDocs) {
+            super(numDocs, getTargetNumGroups(numDocs));
+            assert group100 != null;
+            if (numDocs >= 5000000) {
+                groupIds = group1M;
+            } else if (numDocs >= 500000) {
+                groupIds = group100K;
+            } else {
+                groupIds = group10K;
+            }
+            docsPerGroupBlock = ((float) numDocs) / groupIds.length;
+            reset();
+        }
+
+        @Override
+        public void add(LineFileDocs.LineFileDoc lfd) throws InterruptedException {
+            if (lfd == LineFileDocs.END) {
+                assert docCounter == 0;
+                outputQueue.put(END);
+            }
+            buffer[docCounter++] = lfd;
```

Comment:
Hmm are `LineFileDoc` instances reused (it's unsafe to buffer/hold onto more than one at a time in a thread)?

Reply:
No I don't think they're reused?